### PR TITLE
libnfs: allow building with pthread support

### PR DESCRIPTION
**Pull request by pborzenkov**
###### Description of changes
libnfs is not thread-safe by default. Making it thread-safe requires building
it with pthread support and enabling multi-threading at runtime using
nfs_mt_service_thread_start() function.

The patch allows to optionally build libnfs with multi-threading support.

This is not enabled by default to avoid runtime costs for programs that don't
require libnfs to be thread-safe.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libnfs/default.nix [hunk 1]
```diff
@@ -13,6 +13,17 @@ stdenv.mkDerivation rec {
...
+  patches = [
+    # Fixes 100% CPU usage in multi-threaded mode
+    (fetchpatch {
+      url = "https://github.com/sahlberg/libnfs/commit/34d6fe37e986da5b0ced86cd028a88e482537d5a.patch";
+      sha256 = "sha256-i7mi+TVdkLb4MztT5Ic/Q8XBIWk9lo8v5bNjHOr6LaI=";
+    })
+  ];
+
+  configureFlags = []
+    ++ lib.optional (enableMultithreading) "--enable-pthread";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libnfs/default.nix:25]:
```suggestion
  configureFlags = lib.optional enableMultithreading "--enable-pthread";
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -20406,6 +20406,10 @@ with pkgs;
 
   libnfs = callPackage ../development/libraries/libnfs { };
 
+  libnfs-pthread = callPackage ../development/libraries/libnfs {
+    enableMultithreading = true;
+  };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20411]:
```suggestion
  libnfs-pthread = callPackage ../development/libraries/libnfs { };
```

File: pkgs/development/libraries/libnfs/default.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, autoreconfHook }:
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, enableMultithreading ? false }:
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libnfs/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, enableMultithreading ? true }:
```
Do we need the option at all? If not we might as well remove it.

Comment (pborzenkov) [reply pkgs/development/libraries/libnfs/default.nix:1]: @SuperSandro2000 

Thanks for the review.

> Do we need the option at all? If not we might as well remove it.

Enabling multi-threading by default adds a bunch of conditions here and there and is not enabled by library's author by default for performance reasons. But I think the performance difference, if any, should be negligible. So let me redo the PR and enable `pthread` support by default.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/libraries/libnfs/default.nix [hunk 4]
```diff
@@ -13,6 +13,18 @@ stdenv.mkDerivation rec {
...
+  patches = [
+    # Fixes 100% CPU usage in multi-threaded mode
+    (fetchpatch {
+      url = "https://github.com/sahlberg/libnfs/commit/34d6fe37e986da5b0ced86cd028a88e482537d5a.patch";
+      sha256 = "sha256-i7mi+TVdkLb4MztT5Ic/Q8XBIWk9lo8v5bNjHOr6LaI=";
+    })
+  ];
+
+  configureFlags = [
+    "--enable-pthread"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libnfs/default.nix:25]:
```suggestion
    "--disable-werror"
    "--enable-pthread"
```

File: pkgs/development/libraries/libnfs/default.nix [hunk 5]
```diff
@@ -13,6 +13,18 @@ stdenv.mkDerivation rec {
...
+  ];
+
   NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-error=tautological-compare";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libnfs/default.nix:29]:
```suggestion
```
we can probably drop this now, too

File: pkgs/development/libraries/libnfs/default.nix (hunk 4 above)
Comment (pborzenkov) [reply pkgs/development/libraries/libnfs/default.nix:25]: @SuperSandro2000 I've made a PR to `libnfs` to fix the deprecation warnings on modern macOS: https://github.com/sahlberg/libnfs/pull/403

I'll update this PR to disable `werror` on macOS if the PR is not accepted upstream.

**Pull request merged**